### PR TITLE
Add support for uploading assets to s3

### PR DESCRIPTION
**Pull request by bennettgoble**
Add basic support for uploading release assets to multiple locations: Github Releases and AWS S3 bucket(s). This can be useful in situations, such as with SLVoice, where we have private source but publicly release the distribution artifact.

### CODE REVIEW

**Review by Geenz — APPROVED** [review]
Looks good - though I'd get a second opinion.  I'm not super familiar with our GH actions.

File: README.md [hunk 1]
```diff
@@ -29,7 +29,39 @@ jobs:
...
+
+### Publishing assets to S3
+
+If you want to upload your release artifacts to S3 in addition to github:
+
+```yaml
+steps:
+  - uses: secondlife/action-autobuild-release@v3
+    with:
+      public: true
```
Comment (nat-goodspeed) [inline README.md:45]: This should be very useful! The primary use case, of course, is to post a public autobuild package built from an internal source repo.

But usually, consuming an internal GitHub repo's release asset is exactly when we need `public: false`. Does the `public` parameter affect both the GH release and the S3 bucket? If it does, then the GH release will effectively be unusable -- in which case, we probably wouldn't choose to upload to both destinations.

**Review by nat-goodspeed — COMMENTED** [review]
I do have a question about the `public` action parameter. I'm not familiar with Typescript, so I choose to believe that your code does what you claim.

File: src/upload.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        type: "s3",
+        url: `https://${bucket}.s3.amazonaws.com/${prefix}${name}`,
+      },
+      mermaidGraphFile: path.join(params.artifact.downloadPath, "autobuild-graph.mermaid"),
+    }
+  }
+}
+
+export class GithubArtifactUploader implements ArtifactUploader {
+  constructor(private gh: models.GitHub, private config: Config) {}
```
Comment (phronimos-linden) [inline src/upload.ts:77]: I wasn't familiar with this bit of typescript magic...

**Review by phronimos-linden — APPROVED** [review]
TS stuff seems good 👍

File: README.md (hunk 1 above)
Comment (bennettgoble) [reply README.md:45]: I suppose that parameter could use a bit more documentation. The `public` option in the action only controls the `creds=` parameter placed in the release installation instructions. Whether the github release assets are public or not is wholly based on the publicity of the repository.

Comment (bennettgoble) [reply README.md:45]: Ah, the parameter also controls whether an S3 asset is given the `public-read` or `private` ACL value.

Comment (nat-goodspeed) [reply README.md:45]: `public: false` also controls the form of the URL in the install instructions, but if the presence of an S3 destination means that the install instructions will always use the S3 URL, I guess that's moot.

**Comment by nat-goodspeed** [comment]
Per [Vir's Slack remarks](https://lindenlab.slack.com/archives/CDTHXLF2S/p1707187955404869), it would seem useful to add "upload to Google Drive" capability.

**Pull request merged**